### ISSUE TITLE
Someone put me on the right path to add a pure C++ QQuickItem to React Native

### ISSUE THREAD

**Issue by lucaszanella** [status-im/react-native-desktop#456]
I want to implement something that draws on OpenGL. I already have a subclass of `QQuickItem` which does this, all on C++, no QML required.

As I've been reading, there's a Module Interface, which I presume, is an interface to add new modules, like the one I'm trying to add. As I've read, class ModuleInterface has a virtual method 

```
virtual ViewManager* viewManager() {
        return nullptr;
}
```

I think that for me to add my module, I just have to create a ViewManager and in this manager, I need to implement the following: 

```
QQuickItem* ViewManager::createView(const QVariantMap& properties) {
    QQuickItem* item = new MyOpenGLDrawer();
    return item;
}
```

Is this the right way to implement an item? Because yesterday I edited a lot of the source code in order to add my item just as someone hereadded `Switch`. It became a mess.

**Comment by vkjr** [comment]
@lucaszanella, what component do you want to implement?
If you want to implement component that exists in `react-native` but still doesn't exist in `react-native-desktop` then you can modify code of react-native-desktop.
If you need a component for personal usage that is unique to your project, it shouldn't be part of react-native-desktop codebase and no changes in it required.

**Comment by lucaszanella** [comment]
It's a video player for security cameras, it's not in react-native neither react-native-desktop, im going to implement in both. Can it be implemented as a module? I've found no way in ModuleInterface to return a QQuickItem.

**Comment by lucaszanella** [comment]
> @lucaszanella, what component do you want to implement?
> If you want to implement component that exists in `react-native` but still doesn't exist in `react-native-desktop` then you can modify code of react-native-desktop.
> If you need a component for personal usage that is unique to your project, it shouldn't be part of react-native-desktop codebase and no changes in it required.

I've studied ModuleInterface and I conclude there's no way to specify a QQuickItem to show on screen. Looks like it can only produce non visual APIs

**Comment by lucaszanella** [comment]
It looks like 

https://github.com/status-im/react-native-desktop/blob/321a680f89fd36b8da1494c2555c50138110aaef/ReactQt/runtime/src/moduleinterface.h#L49

was meant to return a ViewManager which in turn would contain my QQuickItem, but it looks it was never finished :(

Can you confirm I'll have to tweak the library?

**Comment by vkjr** [comment]
@lucaszanella, yes looks like you are right. Let me check if I can fix it in nearest days.
In general the class you implement in an external library should inherit not the ModuleInterface, but ViewManager.

**Comment by lucaszanella** [comment]
I think it should inherit both ModuleInterface and ViewManager. That's what I did until I realised the ViewManager inside ModuleInterface was another class.

I'll read the library to see if I can help, but I'm still understanding it. If you come with a fix please tell me, I'm very excited about launching my app in react-native-desktop

**Comment by vkjr** [comment]
ViewManager is a successor of ModuleInterface. It is a base class for all modules that implement visual elements - https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/componentmanagers/viewmanager.h

**Comment by lucaszanella** [comment]
Oh ok. My QQuickItem is ready, so I just need ViewManager to be the class that needs to be implemented to create new modules, so I can add my QQuickItem to it as thus show it to the screen :)

**Comment by vkjr** [comment]
You need to create own `MyOpenGLDrawerManager : public ViewManager` and override couple methods. Similarly to `SwitchManager` that you can find in PR I gave you earlier.

The only thing - I'm not sure if ViewManager will be available the same way as a ModuleInterface when you implement 3rd party lib. This should be checked and maybe tweaked in react-native-desktop side.

**Comment by lucaszanella** [comment]
I already have such manager you mentioned:

https://github.com/lucaszanella/orwell/blob/master/testapp/DesktopSampleApp/orwell-module/orwellmediaplayermanager.cpp#L61

See that it creates my custom object and returns it as a QQuickItem, I think it might work

And yes, it gives an error about viewmanager.h not being found, but I can correct that. I just don't know what to do after it. I'd have to probably tweak more things in the library

**Comment by vkjr** [comment]
@lucaszanella, I created a repo for you that shows how to implement and use custom QQuickItem component:
https://github.com/vkjr/BezierCurveRnd

**Comment by lucaszanella** [comment]
@vkjr thank you so much!!!!! With this I successfully linked my library to libreactnative without even touching the react-native-desktop folder. 

I discovered the additional variables

```
REACT_NATIVE_DESKTOP_EXTERNAL_MODULES_SRC
REACT_NATIVE_DESKTOP_EXTERNAL_PROJECT_DEPS
REACT_NATIVE_DESKTOP_EXTERNAL_MODULES_LIBS
REACT_NATIVE_DESKTOP_EXTERNAL_MODULES_INCLUDE_DIRS

```

Which permits me adding include files and link libraries. I used the to add my library.

Now I just need to fix that npm/yarn thing to test what I did :)

**Comment by naghdy** [comment]
➤ Volodymyr Kozieiev commented:

@lucaszanella, yes looks like you are right. Let me check if I can fix it in nearest days.
In general the class you implement in an external library should inherit not the ModuleInterface, but ViewManager.

**Comment by lucaszanella** [comment]
Hi @vkjr, I'm reading your https://github.com/vkjr/BezierCurveRnd example and you said in the end:

> To make QQuickitem component behave like other react-native views it should inherit this class: https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/reactitem.h and paint() function should be reused.
> 
> To add flexbox properties to QQuickItem component it should have flexbox property. You can look how it set in View.qml and adjust QQuickItem accordingly: https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/qml/ReactView.qml

I already made it inherit ReactItem.h, but what should I implement in the paint() function?

I've looked into https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/qml/ReactView.qml and I couldn't understand what I need to do. I only need to set those variables myself? If My object has those variables it'll have flexbox properties?

**Comment by lucaszanella** [comment]
In order for my custom `QQuickItem` to have Flexbox properties, I did as @vkjr said, based on https://github.com/status-im/react-native-desktop/blob/master/ReactQt/runtime/src/qml/ReactView.qml

```
class OpenGlMaterialQQuickItem: public ReactItem
{
    Q_OBJECT
    Q_PROPERTY(Flexbox* flexbox);
    Q_PROPERTY(ViewManager* viewManager);

    public:
        Flexbox* flexbox;
        ViewManager* viewManager = nullptr;

        OpenGlMaterialQQuickItem()
        {
            flexbox = new Flexbox();
            flexbox->setControl(this);
            flexbox->setViewManager(this->viewManager);
            setFlag(ItemHasContents, true);
        }
```

But my item will not have flexbox properties. Rigth now, react-native-desktop renders on item on top of another, not one below the other.

I found the line in the source code:

`auto childFlexbox = Flexbox::findFlexbox(child);`

which leads me to

`Flexbox* flexbox = control->property("flexbox").value<Flexbox*>();`

which proves that react-native-desktop looks for the property called `"flexbox"` and it's a pointer to `Flexbox`, so I think I'm doing everything all rigth. 

Anyone knows of something that migth be missing?

**Comment by lucaszanella** [comment]
I managed to get it work by adding `READ getFlexbox` in `Q_PROPERTY`. 

To help someone in the future:

```
class OpenGlMaterialQQuickItem: public ReactItem
{
    Q_OBJECT
    Q_PROPERTY(Flexbox* flexbox READ getFlexbox);
    Q_PROPERTY(ViewManager* viewManager);

    public:
        Flexbox* flexbox;
        ViewManager* viewManager = nullptr;
        Flexbox* getFlexbox() {
            return this->flexbox;
        }
        OpenGlMaterialQQuickItem()
        {
            flexbox = new Flexbox();
            flexbox->setControl(this);
            flexbox->setViewManager(this->viewManager);
            setFlag(ItemHasContents, true);
        }
```
